Remove DelayedJob as hard dependency to support Delayed gem

synth opened the pull request:
Delayed gem is a drop in replacement fork for DelayedJob that seems to be more well maintained and have better support Ruby3.

synth committed:
Remove delayed_job as dependency to support delayed gem

Delayed gem is a drop in replacement fork for DelayedJob that seems to be more well maintained and have better support Ruby3.

synth committed:
Add support for Delayed gem by tieing into Railties if possible

synth committed:
Add support for explicit signature

synth committed:
Remove commented line

pabois commented:
Hi @synth 

Thanks for your interest on this gem.
We are waiting for you to set the PR as ready to review, please make sure to check via CodeClimate that the new code does not add any complexity of some sort.

Thank you!

synth commented:
Thanks, will do! We are waiting for our Ruby 3 upgrade to complete to be sure it covers at least our use case, then will submit for review and check CC. Cheers.

synth commented:
@pabois Something has come up in our testing. We came across the random signature generation: https://github.com/noesya/delayed_job_prevent_duplicate/commit/29003a361fb51d31e1d619af4c99bf6193adbc2d. 

Can you share what is the purpose of a randomly generated signature? By default, this would make all jobs that are not `Delayed::PerformableMethod` to be unique and not de-duped. 

We encountered this for jobs that use the [ProgressJob gem](https://github.com/Recognize/progress_job).

070bex424pankaj committed:
Add signature for delayed job enqueue method

SebouChu commented:
@synth We did not find a good solution to differentiate the other jobs so as a temporary solution, we managed them as all different until we find a better way to check the identical ones

synth commented:
What we did is to allow jobs to define their own signature method. 

How do you feel about us adding a `respond_to(:signature)` check and delegation?

pabois commented:
Hello,

Yes, good idea!

yours

synth committed:
Merge pull request #1 from 070bex424pankaj/generate-signature-for-non-performable-method

Add signature for delayed job enqueue method

TwilightCoder committed:
fix: include kw args in signature generation

TwilightCoder committed:
fix: update obsolete variable reference

synth committed:
Merge pull request #2 from TwilightCoder/include-kwargs-when-generating-signature

Include kwargs when generating signature

synth commented:
Quick update, we found some issues on our fork and have patched them. So, once we battle test those updates, we'll submit this PR officially :)

TwilightCoder committed:
feat: Improve signature for class methods

The method args were not being factored previously.

TwilightCoder committed:
refactor: rename method and use logger

pabois commented:
Thanks for the update!

TwilightCoder committed:
fix: complete partial implementation of defensive args handling

TwilightCoder committed:
feat(contd.): account for kwargs in new signature

TwilightCoder committed:
revert/refactor: be defensive when arg matching raises error

synth committed:
Merge pull request #3 from TwilightCoder/improve-signature-for-class-methods

Make signature unique for class methods

TwilightCoder committed:
feat: support custom signatures again for class methods

synth committed:
Merge pull request #4 from TwilightCoder/custom-signature-support-for-class-methods

Support custom signatures for class methods

TwilightCoder committed:
feat: truncate signature when it exceeds column limit

synth committed:
Merge pull request #5 from TwilightCoder/signature-truncation

Signature truncation

070bex424pankaj committed:
Insert ignore duplicate prevention (#6)

* Enhance README and plugin for duplicate job prevention strategies

- Clarified the purpose and functionality in README.
- Added detailed descriptions for duplicate prevention strategies: :validation, :insert_ignore, and :on_duplicate_key.
- Updated the plugin to support new strategies and improved the save method for handling duplicates.
- Ensured unique index creation in migration template for the signature column.

* fix: improve error handling in plugin

* fix: implement atomic signature locking for job re-enqueueing

* fix: update signature locking implementation to use self.prepended pattern
